lisp/bazel-build.el

```diff
@@ -25,38 +25,109 @@
+
+(defvar bazel-query-args '("query" "--noblock_for_lock")
+  "Bazel query subcommand and arguments.")
+
+(defvar bazel-run-commands-from-root nil
+  "A flag indicating to run commands from the workspace root.")
+
 (defun bazel-build--run-bazel-command (command target)
   "Run Bazel tool with given COMMAND, e.g. build or run, on the given TARGET."
-  (compile
-   (mapconcat #'shell-quote-argument (list "bazel" command target) " ")))
-
-(defun bazel-build--read-target (prompt)
-  "Read a Bazel build target from the minibuffer.  PROMPT is a read-only prompt."
-  (let* ((file-name (buffer-file-name))
-         (workspace-root
-          (or (bazel-util-workspace-root file-name)
-              (user-error "Not in a Bazel workspace.  No WORKSPACE file found")))
-         (package-name
-          (or (bazel-util-package-name file-name workspace-root)
-              (user-error "Not in a Bazel package.  No BUILD file found")))
-         (initial-input (concat "//" package-name)))
-    (read-string prompt initial-input)))
+  (let ((command
+         (if bazel-run-commands-from-root
+             (let ((root-dir (or (bazel-util-workspace-root
+                                  (or (buffer-file-name)
+                                      (and dired-directory (directory-file-name dired-directory))))
+                                 (user-error "Could not find workspace."))))
+               (format "cd %s && %s %s %s" root-dir
+                       bazel-command command target))
+           (mapconcat #'shell-quote-argument (list bazel-command command target) " "))))
+    (compile command)))
+
+(defun bazel-build--process (&rest args)
+  "Run a bazel subcommand. Return stdout as string."
+  ;; Note: The current working directory of the subprocess is set to the current
+  ;; buffer's value of default-directory.
+  (let ((stderr-file (make-temp-file "bazel-stderr")))
+    (unwind-protect
+        (string-trim-right
+         (with-output-to-string
+           (let ((status
+                  (apply #'process-file bazel-command
+                         nil (list standard-output stderr-file) nil args)))
+             (if (/= status 0)
+                 (with-temp-buffer
+                   (insert-file-contents stderr-file)
+                   (error "Error running command: %s" (buffer-string)))))))
+      (if (file-exists-p stderr-file)
+	  (delete-file stderr-file)))))
+
+(defun bazel-query (&rest args)
+  "Run a bazel query subcommand. Return stdout as string."
+  (apply #'bazel-build--process (append bazel-query-args args)))
+
+(defun bazel-build--target-for-file (filename)
+  "Get the list of targets which includes the given filename."
+  (let* ((default-directory (file-name-directory filename))
+         ;; Resolve label for file with bazel query.
+         (fullname (bazel-query (file-name-nondirectory filename))))
+    ;; Produce target the file-label is in using bazel query.
+    (let ((cmd (format "attr('srcs', %s, %s:*)" fullname
```

Comment:
I'd suggest `same_pkg_direct_rdeps` - that will find direct consumers of the file even if they don't use the `srcs` attribute.